From dd7a24068f037d0ff895ab94b1c0a8ed4963c503 Mon Sep 17 00:00:00 2001
From: Li Jun <jun.li@freescale.com>
Date: Fri, 23 Jan 2015 18:35:35 +0800
Subject: [PATCH 1417/1543] MLK-10132-3 usb: chipidea: udc: do not enter low
 power mode if vbus on

commit 9b209436ec923eca8e91510d9bca332a50360328 from
git://git.freescale.com/imx/linux-2.6-imx.git

This patch is to prevent usb entering low power mode if vbus is on even gadget
driver is not binded, by holding the PM count of ci->dev.
So, there are 3 pm usage_count status:
- ci->dev: 1 ci->gadget.dev: 1
  Device mode with gadget driver binded and vbus on.
- ci->dev: 1 ci->gadget.dev: 0
  USB vbus on but gadget driver not binded.
- ci->dev: 0 ci->gadget.dev: 1
  USB OTG FSM is in a_peripheral mode.
Above 2 device's pm usage_count hold by ci otg(ci->dev) and usb gadget
(ci->gadget.dev).

Signed-off-by: Li Jun <jun.li@freescale.com>
---
 drivers/usb/chipidea/udc.c |    9 +++++++--
 1 files changed, 7 insertions(+), 2 deletions(-)

diff --git a/drivers/usb/chipidea/udc.c b/drivers/usb/chipidea/udc.c
index 4743c6e..8a0724d 100644
--- a/drivers/usb/chipidea/udc.c
+++ b/drivers/usb/chipidea/udc.c
@@ -1865,8 +1865,10 @@ int ci_usb_charger_connect(struct ci_hdrc *ci, int is_active)
 {
 	int ret = 0;
 
+	if (is_active)
+		pm_runtime_get_sync(ci->dev);
+
 	if (ci->platdata->notify_event) {
-		pm_runtime_get_sync(&ci->gadget.dev);
 		if (is_active)
 			hw_write(ci, OP_USBCMD, USBCMD_RS, 0);
 
@@ -1881,8 +1883,11 @@ int ci_usb_charger_connect(struct ci_hdrc *ci, int is_active)
 			/* Pull down dp */
 			hw_write(ci, OP_USBCMD, USBCMD_RS, 0);
 		}
-		pm_runtime_put_sync(&ci->gadget.dev);
 	}
+
+	if (!is_active)
+		pm_runtime_put_sync(ci->dev);
+
 	return ret;
 }
 
-- 
1.7.5.4

